### ISSUE TITLE
Theoretical and practical documentation for using transform matrices in Godot

### ISSUE THREAD

**Issue by tagcup** [godotengine/godot#17951]
I can't reply to #17949 either because it's closed and gives me "You can't comment at this time. "...

So, anyway, replying to @reduz [there](https://github.com/godotengine/godot/issues/17949#issuecomment-378387595):

> Working with transforms directly is clearly difficult for a lot of people. We should really document this better.

 Yeah, I completely agree, we ideally need a document that covers 1) all practical use cases in a shallow language 2) a mathematical text for those who'd like to understand the shallow language.

I mentioned this multiple occasions, I do have an incomplete write up (it's a newly created rst file, so it doesn't conflict with any existing docs.), and it's been sitting like that for a while, I can clean it up a little and dump whatever I currently have as a PR this weekend. I can send further PRs toward making it a complete guide. Would that be acceptable?

About what I currently have:

The math part currently includes a very brief introduction to Lie groups and Lie algebras, representation theory , a review of SO(2)/U(1) isomorphism (real vs complex representation of), and finally SO(3)/SU(2)/Spin(3) (real vs complex), and there's also a section about parametrization of the group elements (axis-angle for general Lie groups, and Euler angles for SO(3)). I also have a bare-bones section about SLERP.

I know, it probably sounds too technical due to many mathematical terms, but I tried write it in a totally non-technical and self-contained way, such that someone who only knows high school math can understand it.

The practical use text is a bit weak right now; I have a couple of tables along the lines "use this if you want to achieve this", and a table which clarifies representation vs parametrization, but we can hopefully improve that based on feedback and issue reports like this.

So let me know if this makes sense to you or not.

**Comment by tagcup** [comment]
I don't plan on adding a linear algebra primer, but I would add a section about reference frames, how matrices and vectors transform as you switch the frame, and finally about the 3 reference frames used in Godot (global, parent-local, object-local)

**Comment by akien-mga** [comment]
> I can't reply to #17949 either because it's closed and gives me "You can't comment at this time. "...

That's super weird, I don't understand why this would happen :/ The issue is only closed, not locked, so you should be able to continue writing there in theory. Well, let's continue discussion here then.

**Comment by tagcup** [comment]
and of course there should be something about scaling and translations, and why exactly uniform scaling is necessary in Godot, and also something about why we can't do polar decomposition when negative scale/reflections are involved, because I remember seeing issues about those

**Comment by tagcup** [comment]
@akien-mga Yeah, that surprised me as well, and I'm also not exactly sure how the user blocking thing works in Github (maybe it's related to that), I didn't even know about it until today

**Comment by cbscribe** [comment]
This would be a great addition to the "Math" section. The beginner vector doc was updated a while ago but it gets very sparse when you get past that. The existing "Matrices and Transforms" section needs work, for example.

I would be happy to review and help with your PR. I think it might be something that we split into multiple docs so readers can go as far down the rabbit hole as they need to. The more esoteric, less frequently encountered issues should be later in the doc.

**Comment by akien-mga** [comment]
Looks like @reduz went ahead and wrote something about transforms already: https://godot.readthedocs.io/en/latest/tutorials/3d/using_transforms.html

It would still be quite interesting to have this reviewed and completed based on what you've been working on on your side.

**Comment by tagcup** [comment]
That looks nice, we can finally point to something in issues!

What I have probably doesn't fit into it, but maybe parts of it could be merged. I'll just submit it as a separate WIP doc, and maybe we can work it from here.

Some quick comments on @reduz's write up

In Introduction
X is the Right axis, Y is the Up axis and Z is the Forward axis.

This can be a little difficult to parse because while this description is true if you're looking at a coordinate system from the camera (and forward means to the camera), when you're the object (like in FPS), X is to the left.

In the Axis Order section:
1. the axis order is always *fixed* by a convention, it's hard-coded to be XYZ in Godot for example, so I wouldn't even mentioned about how the convention is fixed
2. with that ordering in mind, a sequence of Euler rotations to realize different  rotations
3. gives a picture of gimbal lock but it's not explained and an actual gimbal lock isn't demonstrated in what follows).

In Interpolation section:
1. It's actually fine to have the first and last axes to be aligned (as long as two neighboring axes don't align, it's OK)
2. We should definitely add something about SLERP (with quaternions, since SLERP in matrix representation isn't implemented; we can actually add it but it'll be slower because a matrix is more general than a quaternion which translates to more CPU instructions)

I do have entire sections about Euler angles, what a gimbal lock is, why Euler angles isn't suitable for interpolation, what SLERP, quaternion is, ... so I guess their details can be deferred to my write up

**Comment by tagcup** [comment]
@cbscribe Thanks! The write up is fine in places but poor/very draftish in other places. I hope you'd be fine with a WIP. My hope is to get some feedback on it and expand it.

**Comment by reduz** [comment]
I didn' t see you were working on this, I went ahead as I was talking to users online and requested it today. Apologies as I probably made you waste some time.

> This can be a little difficult to parse because while this description is true if you're looking at a coordinate system from the camera (and forward means to the camera), when you're the object (like in FPS), X is to the left.

You are right, I think it should be probably explained in the context of the camera and object, as is it s not clear.

> the axis order is always fixed by a convention, it's hard-coded to be XYZ in Godot for example

Ah, totally forgot to note this.. 

I'm not quite sure how to explain gimbal lock properly, you are right also that it' s neighboring angles that cause problem. SLERP also makes sense, 

If you want to add sections to it via PR, feel free, otherwise I can add these myself. I am mostly not sure how to explain gimbal lock in a way that is more obvious though.

**Comment by tagcup** [comment]
Not at all, there's maybe 1% overlap with what you wrote and what I wrote. Like I mentioned, I basically have 2 docs: theory and practice, and only theory one (which reads like a math text) is more or less complete. So if anything, it's good for me, because I can instead add practical, non-math-heavy stuff to it rather than trying to make mine complete.

BTW, it's actually YXZ it seems. I added XYZ to C++ at some point but it's not exposed so we can probably forget about it.

I explained gimbal lock in around 600 words, but it's a bit mathematical. I managed to keep it short thanks to previous sections on the Lie group SO(3). I had to explain what 3-torus and real projective space RP3 in words, explained their topological differences. I finally listed the singularities in both axis-angle and euler angle parameterization (north/south pole singularity vs gimbal lock) and concluded that smooth change in one parameter space cannot always be mapped onto the other.

I also have a derivation of SLERP (as a torque-free and reference frame independent interpretation between to elements of SO(3)), so you should be able to link to those if you like.

Last but not least, I'm pretty sure quaternions will make perfect sense to the reader.

But it's all abstract I guess and what I'm missing is of course a practical demonstration of these...


So if you had something in mind for those, please add them. I'll submit only the theory part this weekend, and once it's in place, I can start contributing to the practice part you started.

**Comment by tagcup** [comment]
BTW I added sphinx.ext.mathjax extension to the docs to type the equations (I have so many of them), I hope that's fine

**Comment by vnen** [comment]
@tagcup there's the sphinx.ext.imgmath extension, which is enabled in the docs already (godotengine/godot-docs#620).

**Comment by tagcup** [comment]
Frankly it sucks, it looks like Wikipedia math images from 20 years ago and changing an equation requires adding removing new files. That was necessary before mathjax was a thing. so I prefer to use mathjax

**Comment by tagcup** [comment]
The number of equations that I have is probably measured in hundreds so the problem is quite magnified

**Comment by akien-mga** [comment]
@tagcup It's fine to change the math extension we use is mathjax is better, we should just port the current code using imgmath to avoid having two extensions (it doesn't hurt, but it's better to enforce one for consistency and reduce dependencies).

I don't understand why you say that imgmath requires adding new files though? This code looks like typical LaTeX markup to me: https://github.com/godotengine/godot-docs/blob/master/tutorials/math/vector_math.rst#dot-product
Output: https://godot.readthedocs.io/en/3.0/tutorials/math/vector_math.html#dot-product (I agree it's a bit old school :))

**Comment by akien-mga** [comment]
> The number of equations that I have is probably measured in hundreds so the problem is quite magnified

I'm quite looking forward to reading all this and freshen up a bit my math-fu, I've grown rusty :D

**Comment by tagcup** [comment]
I mean each equation is rendered and stored as a new raster image file. So if you have 100 equations you'll need to commit those 100 files, and the file is the hash of the equation so whenever you change one you need to find the old file, remove it from git, and commit the new file which doesn't scale with the number of equations and is a terrible experience. It also looks blurry and doesn't scale on screen either.

**Comment by tagcup** [comment]
I dealt with a similar system like 15 years ago and it sucked. Mathjax was a revolution in that regard

**Comment by akien-mga** [comment]
> I mean each equation is rendered and stored as a new raster image file. So if you have 100 equations you'll need to commit those 100 files, and the file is the hash of the equation so whenever you change one you need to find the old file, remove it from git, and commit the new file which doesn't scale with the number of equations and is a terrible experience.

Ah I see what you mean. It's not a problem in our case as we only commit source files, Sphinx and ReadTheDocs take care of building the HTML (and here the PNGs) for each commit.

Still, if only for the better visual output I'm happy to switch to mathjax.

**Comment by tagcup** [comment]
An alternative is katex but it doesn't look like they have a plug-in for it

**Comment by tagcup** [comment]
Ah, I see. So they probably remove all the old files and do a fresh build every time.

Yeah, apart from the visuals, there's also the fact that an equation is supposed to be a text (which you can select/copy), not an image.

**Comment by tagcup** [comment]
> I'm quite looking forward to reading all this and freshen up a bit my math-fu, I've grown rusty :D

It'd be great to have some feedback, so please do tell about it if you read it this weekend. It'll hopefully make quaternions and SLERP boring and non-magical for any reader.
